### PR TITLE
 Implement get all installed extensions 

### LINKED ISSUE #5899: Show namespaces for viz, cni and others
### What problem are you trying to solve?

Currently in the Service mesh details table it shows some details but would be great to add all other namespaces information to it.

### How should the problem be solved?

Screenshot 2021-03-14 at 18 36 49

It would be nice to add the other namespaces to this list if they are installed

### PR DESCRIPTION
**Pull request by sannimichaelse**
Show All Installed Extensions

Currently, the Service mesh details table shows some details but would be great to add installed extensions to it

Add a new section called `Installed Extensions`, Modify get extension endpoint to List all installed extensions as well support getting single ones

Once merged, The installed extensions section should be list all installed extensions

Fixes #5899

Signed-off-by: Sanni Michael <sannimichaelse@gmail.com>

Previous Look

image

New Look

image

### CODE REVIEW

**Comment by Pothulapati** [comment]
Hey @sannimichaelse 👋🏼 

So, I just tried testing out this PR but couldn't actually see the extensions being displayed.

image
As you can see, I have both `linkerd-viz` and `linkerd-smi` installed but still can't see them in the list. I can take a look later today on what's going wrong, if you are not able to find the issue. :)

**Comment by sannimichaelse** [comment]
Hey @Pothulapati thanks for checking it out. Found the issue and fixed it already. I also squashed all the commits to keep the commit history clean. You can try to test again

image

File: web/app/js/components/ServiceMesh.jsx
Comment (mateiidavid) [inline web/app/js/components/ServiceMesh.jsx:127]: I think it'd be easier here to just return the resulting array from map.

We declare an empty array on L123, if the state is not empty (and we have a list of extensions) then we map through and change an extension to be its associated metadata. If the array is empty then mapping through it will still give an empty array, right?

```suggestion
    const { extensions } = this.state;
    return extensions.extensions.map(extension => extension.metadata);
```

also, any way we can avoid doing `extensions.extensions` and instead do just `extensions.map`?

File: web/app/js/components/util/ApiHelpers.jsx
Comment (mateiidavid) [inline web/app/js/components/util/ApiHelpers.jsx:154]: I think the name might be a bit misleading here. It seems that when we don't add a query param to the `GET /api/extension` path we return all of the extensions. I'm wondering if this shouldn't be its own handler and its own call.

Perhaps we might also be able to rename the path so the get call goes to `/api/extensions/` in which case `/api/extensions?extension_name=viz` might also make sense. Let me know if this makes sense and what your thoughts around this are.

File: web/srv/api_handlers.go
Comment (mateiidavid) [inline web/srv/api_handlers.go:459]: wdyt about reducing the nesting here with an early return?

```suggestion
	if extensionName == "" {
		extensions, err := h.k8sAPI.GetAllNamespacesWithExtensionLabel(ctx)
		if err != nil {
			renderJSONError(w, err, http.StatusInternalServerError)
			return
		}
		
		resp["extensions"] = extensions
		renderJSON(w, resp)
		return 
	} 
	
	ns, err := h.k8sAPI.GetNamespaceWithExtensionLabel(ctx, extensionName)
	if err != nil && strings.HasPrefix(err.Error(), "could not find") {
		renderJSON(w, resp)
		return
	} else if err != nil {
		renderJSONError(w, err, http.StatusInternalServerError)
		return
	}
	resp["extensionName"] = ns.GetLabels()[k8s.LinkerdExtensionLabel]
	resp["namespace"] = ns.Name
	
	renderJSON(w, resp)
```

**Review by mateiidavid — COMMENTED** [review]
Thank you @sannimichaelse! I built & deployed your changes to my local k3d cluster and it seems the issue with extensions not being rendered has been fixed,  can see all of my installed extensions :) 

I left some general comments on the code, curious what you think.

Also, a thing that stood out to me visually (and I noticed this in @Pothulapati's screenshot too), the heading seems to be wrapped. Is that something that we want? 

image

File: web/app/js/components/ServiceMesh.jsx
Comment (sannimichaelse) [reply web/app/js/components/ServiceMesh.jsx:127]: So if the thing is, the test passes a different value to the service mesh component on component did mount. I noticed it was passing 

```
{ metrics: [] }
```

So the `extensions.extensions` resulted in undefined. And that's why the test was failing initially. I had to put a check to ensure it exists before mapping it

File: web/app/js/components/util/ApiHelpers.jsx
Comment (sannimichaelse) [reply web/app/js/components/util/ApiHelpers.jsx:154]: Yea thought about this too but from the comment from the issue it was stated that the api/extension route should be modified to accommodate getting all and single extensions. You can check more info on that on the original issue #5899

File: web/srv/api_handlers.go
Comment (sannimichaelse) [reply web/srv/api_handlers.go:459]: Yea this looks good, would update the PR. 

Thanks @mateiidavid

File: web/app/js/components/ServiceMesh.jsx
Comment (sannimichaelse) [reply web/app/js/components/ServiceMesh.jsx:127]: For the wrapped heading - would check it out again and see why its like that

File: go.sum
Comment (dadjeibaah) [inline go.sum:240]: Are these changes related to this PR? If not I would back this out.

File: web/srv/api_handlers.go
Comment (dadjeibaah) [inline web/srv/api_handlers.go:443]: It looks like when we get this list of extensions from the k8s API, we send all the API info to the frontend. 
image

The only info we need in the frontend is the extension name and the namespace. I would extract that from what we receive from the k8s API and send back a list of objects that contain the extension's name and its namespace.

File: web/app/js/components/util/ApiHelpers.jsx
Comment (dadjeibaah) [reply web/app/js/components/util/ApiHelpers.jsx:154]: I agree with @mateiidavid on renaming the endpoint. I think now that we are adding the capability to return a list of extensions, it makes more sense to rename the endpoint to `/api/extensions` as well as the method name. On the other hand, I think it would be ok to use one single handler for both cases since the info we dealing with are similar.

File: web/app/js/components/ServiceMesh.jsx
Comment (dadjeibaah) [inline web/app/js/components/ServiceMesh.jsx:125]: If we send back only the info we need from the backend, there shouldn't be a need to go through this extra step of filtering.

**Review by dadjeibaah — COMMENTED** [review]
This is looking good @sannimichaelse! Thanks for submitting this PR! I left a few comments for your consideration.

File: web/app/js/components/ServiceMesh.jsx
Comment (sannimichaelse) [reply web/app/js/components/ServiceMesh.jsx:125]: Thanks for feedbacks, would check them out

**Comment by dadjeibaah** [comment]
@sannimichaelse, checking in to see if you have any questions on this PR.

**Comment by sannimichaelse** [comment]
> @sannimichaelse, checking in to see if you have any questions on this PR.

Hi @dadjeibaah, no questions really - been busy with final exams - rounding up this week. Should get to it this week. Sorry for any inconvenience

File: web/srv/api_handlers.go
Comment (mateiidavid) [inline web/srv/api_handlers.go:449]: I think we're missing the `namespace` here? looking at issue, it seems the namespace should be present for the extension. Is this something that we want to display?

cc. @dadjeibaah

File: web/srv/api_handlers.go
Comment (mateiidavid) [inline web/srv/api_handlers.go:457]: this is a bit nitpicky but I think this block can be a bit confusing w.r.t to the data that's being copied. wdyt about renaming some of these variables to make it a bit clearer to understand what's going on?

for example (feel free to swap with your own)
```suggestion
			for _, installedExtension := range installedExtensions {
			  extension.Name = installedExtension.ObjectMeta.Name
```

It's pretty nitpicky so TIOLI :)

File: web/srv/api_handlers.go
Comment (mateiidavid) [inline web/srv/api_handlers.go:451]: What is the purpose of the `UID` here?

Comment (sannimichaelse) [reply web/srv/api_handlers.go:451]: On the react side, the table needs a unique key while iterating through the values - that's why I decided to add the UID instead of generating a new id since its part of the info

**Review by dadjeibaah — COMMENTED** [review]
This is looking great @sannimichaelse. I've left a bunch of comments for you to consider.

**Comment by dadjeibaah** [comment]
@sannimichaelse, it looks like you'd also need to rebase your PR against the current main branch since that has some fixes that'll make the integration tests pass.

**Review by dadjeibaah — APPROVED** [review]
Thanks for making those updates. The changes look great and work well. Things look good on my side but It'd be good to get another pair of eyes on the PR as well. 📦 🚢

File: web/app/js/components/ServiceMesh.jsx
Comment (mateiidavid) [inline web/app/js/components/ServiceMesh.jsx:41]: Can we please also show the namespace here as previously discussed?

**Review by mateiidavid — CHANGES_REQUESTED** [review]
@sannimichaelse this looks really good! The code's looking nice, however, we're still missing the namespace in the table :( would you mind addressing that? after it should be good to merge.

image

**Review by mateiidavid — APPROVED** [review]
Thanks @sannimichaelse for your effort! PR looks good :)

**Comment by sannimichaelse** [comment]
I agree to the DCO for all the commits in this PR

**Pull request merged**
